src/consul.rs

```diff
@@ -0,0 +1,308 @@
+                .send()
+                .await?
+                .error_for_status()?;
+            let body = resp.text().await?;
+
+            // If the lock is acquired, the response body will be "true"
+            if body.starts_with("true") {
+                println!("Acquired Consul lock");
+                return Ok(());
+            }
+            sleep(Duration::from_millis(100)).await;
+        }
+    }
+
+    /// Drop a lock
+    pub async fn drop_lock(&self) -> Result<()> {
+        let mut lock_url = self.kv_api_base_url.join("consul_lock")?;
+        if let Some(dc) = &self.datacenter {
+            lock_url.query_pairs_mut().append_pair("dc", dc);
+        }
+        self.http_client
+            .delete(lock_url)
+            .send()
+            .await?
+            .error_for_status()?;
+        println!("Dropped Consul lock");
+        Ok(())
+    }
+
+    /// Retrieves a list of all registered services and parses their tags into DnsTag
+    pub async fn fetch_service_tags(
+        &self,
+        consul_index: &mut Option<u64>,
+    ) -> Result<Vec<DnsRecord>, anyhow::Error> {
+        let mut services_url = self.catalog_api_base_url.join("services")?;
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            services_url.query_pairs_mut().append_pair("dc", dc);
+        }
+
+        if let Some(index) = consul_index {
+            services_url
+                .query_pairs_mut()
+                .append_pair("index", &index.to_string());
+            services_url.query_pairs_mut().append_pair("wait", "100s");
+        }
+
+        // Add a filter to only match "normal" Consul services
+        services_url.query_pairs_mut().append_pair(
+            "filter",
+            r#"ServiceKind == "" and ServiceTags contains "external-dns.enable=true""#,
+        );
+
+        let response = self.http_client.get(services_url).send().await?;
+
+        if response.status().is_success() {
+            *consul_index = response
+                .headers()
+                .get("X-Consul-Index")
+                .and_then(|value| value.to_str().ok())
+                .and_then(|value| value.parse().ok());
+        }
+
+        let records = response
+            .error_for_status()?
+            .json::<HashMap<String, Vec<String>>>()
+            .await?;
+
+        let dns_tags = records
+            .into_iter()
+            .flat_map(|(_service_name, tags)| parse_dns_tags(tags))
+            .collect();
+
+        Ok(dns_tags)
+    }
+
+    /// Stores a single DNS record in Consul.
+    /// This function fetches the current state of DNS records, updates it with the new record,
+    /// and then re-stores the updated state back into Consul.
+    pub async fn store_dns_record(
+        &self,
+        provider_record_id: String,
+        dns_record: &DnsRecord,
+    ) -> Result<()> {
+        let mut records = self.fetch_all_dns_records().await?;
```

Comment:
Reading the state for every record created or deleted feels rather heavy-weight.  Since only one instance is allowed to modify the state while the lock is held we can instead just fetch it once and store it as an optional field on `Self`.